File: lc-python/src/solutions/lc_0226_invert_binary_tree.py
import unittest
from typing import Optional

from shared.tree import TreeNode, list_tree_to_tree, tree_to_list


def invert_tree0(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """

    # traverse the tree using depth-first traversal
    # when visiting a node, swap its leaves
    def dfs(node: Optional[TreeNode]) -> Optional[TreeNode]:
        # base case
        if node is None:
            return None

        # swap tree nodes
        temp = node.left
        node.left = node.right
        node.right = temp

        # traverse
        dfs(node.left)
        dfs(node.right)

        return node

    return dfs(root)


def invert_tree1(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """
    s = [root]
    while s:
        n = s.pop()

        # base case
        if n is None:
            continue

        # swap
        tmp = n.left
        n.left = n.right
        n.right = tmp

        # traverse
        s.append(n.left)
        s.append(n.right)

    return root


def invert_tree2(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """
    Time complexity: O(n)
    Space complexity: O(n)
    """
    q = [root]
    while q:
        n = q.pop(0)

        # base case
        if n is None:
            continue

        # swap
        tmp = n.left
        n.left = n.right
        n.right = tmp

        # traverse
        q.append(n.left)
        q.append(n.right)

    return root


class TestInvertBinaryTree(unittest.TestCase):

    @staticmethod
    def parse_input():
        data = [
            [[4, 2, 7, 1, 3, 6, 9], [4, 7, 2, 9, 6, 3, 1]],
            [[2, 1, 3], [2, 3, 1]],
            [[], []],
        ]
        return data

    def test_invert_binary_tree0(self):
        data = self.parse_input()

        for tc in data:
            h = tree_to_list(invert_tree0(list_tree_to_tree(tc[0])))
            self.assertEqual(h, tc[1])

    def test_invert_binary_tree1(self):
        data = self.parse_input()

        for tc in data:
            h = tree_to_list(invert_tree1(list_tree_to_tree(tc[0])))
            self.assertEqual(h, tc[1])

    def test_invert_binary_tree2(self):
        data = self.parse_input()

        for tc in data:
            h = tree_to_list(invert_tree2(list_tree_to_tree(tc[0])))
            self.assertEqual(h, tc[1])


if __name__ == "__main__":
    unittest.main()

File: lc-typescript/src/solutions/lc_0226_invert_binary_tree.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";

import { TreeNode, array_to_tree, tree_to_array } from "../shared/utils.ts";

function invert_tree0(root: TreeNode | null): TreeNode | null {
    const stack = [root];
    while (stack.length > 0) {
        const n = stack.pop();
        if (n == null) {
            continue;
        }

        // swap
        const t = n.left;
        n.left = n.right;
        n.right = t;

        // traverse
        stack.push(n.left);
        stack.push(n.right);
    }

    return root;
};

function invert_tree1(root: TreeNode | null): TreeNode | null {
    function dfs(node: TreeNode | null) {
        if (node == null) {
            return;
        }

        // swap
        const t = node.left;
        node.left = node.right;
        node.right = t;

        // traverse
        dfs(node.left);
        dfs(node.right);
    }

    dfs(root);

    return root;
}

Deno.test("0226 Invert Binary Tree", async (t) => {
    const test_cases = [
        [[4, 2, 7, 1, 3, 6, 9], [4, 7, 2, 9, 6, 3, 1]],
        [[2, 1, 3], [2, 3, 1]],
        [[], []],
    ];

    await t.step(`Iterative DFS`, () => {
        for (let i = 0; i < test_cases.length; i += 1) {
            const test_case = test_cases[i];

            const tree1 = array_to_tree(test_case[0]);
            const tree3 = invert_tree0(tree1);

            const actual = tree_to_array(tree3);
            const expected = test_case[1];

            assertEquals(actual, expected);
        }
    });

    await t.step(`Recursive DFS`, () => {
        for (let i = 0; i < test_cases.length; i += 1) {
            const test_case = test_cases[i];

            const tree1 = array_to_tree(test_case[0]);
            const tree3 = invert_tree1(tree1);

            const actual = tree_to_array(tree3);
            const expected = test_case[1];

            assertEquals(actual, expected);
        }
    });

});
